### PR TITLE
Make materials not issue a warning multiple times

### PR DESCRIPTION
**Pull request by YaqiWang**
Closes #26173.

### CODE REVIEW

File: framework/include/materials/Material.h [hunk 1]
```diff
@@ -217,6 +217,8 @@ class Material : public MaterialBase, public Coupleable, public MaterialProperty
   /// resolve all optional properties
   virtual void resolveOptionalProperties() override;
 
+  virtual bool shouldWarn() const override { return !_bnd || boundaryRestricted(); }
```
Comment (lindsayad) [inline framework/include/materials/Material.h:220]: The concern I have with this is what if a developer warns for some condition that is only triggered in a face or boundary evaluation?

Comment (YaqiWang) [reply framework/include/materials/Material.h:220]: I think that would be a rare case compared with this. A developer can override this function again in his material. Note that the logic here is to warn on element and boundary materials except face and neighbor.

**Comment by roystgnr** [comment]
Not sure I understand what's going on here.  We're emitting warnings from a `Material`, but for some reason (what?) the same warning is getting emitted both on internal and on boundary evaluations, and you're turning off the boundary evaluation emissions?

Is that the best way to solve the problem?  Surely the behavior we want is "emit a warning iff it hasn't already been emitted"?  E.g. in libMesh our `libmesh_warning()` wraps around `libmesh_do_once()`, which uses a block-local `static bool did_this_already = false`, testing that it's still false (and setting it to true) before continuing.  Not perfect (in theory you can still get the same warning twice if two threads hit it at the *exact* same time) but probably more robust than hoping that a warning isn't evaluation-location-dependent.

**Comment by YaqiWang** [comment]
> Not sure I understand what's going on here. We're emitting warnings from a `Material`, but for some reason (what?) the same warning is getting emitted both on internal and on boundary evaluations, and you're turning off the boundary evaluation emissions?
> 
> Is that the best way to solve the problem? Surely the behavior we want is "emit a warning iff it hasn't already been emitted"? E.g. in libMesh our `libmesh_warning()` wraps around `libmesh_do_once()`, which uses a block-local `static bool did_this_already = false`, testing that it's still false (and setting it to true) before continuing. Not perfect (in theory you can still get the same warning twice if two threads hit it at the _exact_ same time) but probably more robust than hoping that a warning isn't evaluation-location-dependent.

The multiple warnings are from the three constructions of the same material on element, face and neighbor face by MOOSE in `FEProblemBase::addMaterial`. Typically one warning is sufficient. I did not try but I guess wrapping the mooseWarning with mooseDoOnce won't help because there are three places in memory.

**Comment by roystgnr** [comment]
> I guess wrapping the mooseWarning with mooseDoOnce won't help because there are three places in memory.

Give it a try?  A function-local `static bool` doesn't get created once per object instantiation, it gets created once per function instantiation.  If you create three different `FooMaterial` objects and the class has a `mooseDoOnce(mooseWarning(etc))` in `FooMaterial::bar()` you should only get at most one warning out of it.

The only catch is that you have to know just how many times your function is getting instantiated: e.g. if you put the `mooseDoOnce` inside `Material` then you're making a mistake because a `FooMaterial` warning could then suppress an unrelated `BazMaterial` warning; or if you put the `mooseDoOnce` inside `FooMaterial::bar<T>` then you might get the same warning repeated for three different type `T`.

**Comment by YaqiWang** [comment]
I think I will just override the two functions in my derived classes. Because all my derived classes share the same base class and they should always be block restricted, it makes sense to do the overriding there and the change can be done easily. I am not sure whether you want to close the issue though. Thanks.

**Pull request closed without merging**